src/main/java/org/opensearch/knn/index/memory/SharedIndexStateManager.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.memory;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.knn.jni.JNIService;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Class manages allocations that can be shared between native indices. No locking is required.
+ * Once a caller obtain an instance of a {@link org.opensearch.knn.index.memory.SharedIndexState}, it is guaranteed to
+ * be valid until it is returned. {@link org.opensearch.knn.index.memory.SharedIndexState} are reference counted
+ * internally. Once the reference count goes to 0, it will be freed.
+ */
+@Log4j2
+class SharedIndexStateManager {
+    // Map storing the shared index state with key being the modelId.
+    private final ConcurrentHashMap<String, SharedIndexStateEntry> sharedIndexStateCache;
+    private final ReadWriteLock readWriteLock;
+
+    private static SharedIndexStateManager INSTANCE;
+
+    // TODO: Going to refactor away from doing this in the future. For now, keeping for simplicity.
+    public static synchronized SharedIndexStateManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new SharedIndexStateManager();
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Constructor
+     */
+    @VisibleForTesting
+    SharedIndexStateManager() {
+        this.sharedIndexStateCache = new ConcurrentHashMap<>();
+        this.readWriteLock = new ReentrantReadWriteLock();
+    }
+
+    /**
+     * Return a {@link SharedIndexState} associated with the key. If no value exists, it will attempt to create it.
+     * Once returned, the {@link SharedIndexState} will be valid until
+     * {@link SharedIndexStateManager#release(SharedIndexState)} is called. Caller must ensure that this is
+     * called after it is done using it.
+     *
+     * In order to create the shared state, it will use the indexAddress passed in to create the shared state from
+     * using {@link org.opensearch.knn.jni.JNIService#initSharedIndexState(long, KNNEngine)}.
+     *
+     * @param indexAddress Address of index to initialize the shared state from
+     * @param knnEngine engine index belongs to
+     * @return ShareModelContext
+     */
+    public SharedIndexState get(long indexAddress, String modelId, KNNEngine knnEngine) {
+        this.readWriteLock.readLock().lock();
```

Comment:
does this read lock prevents the release to be happening? what will happen if one thread has read the index state and at the same time a release is called to remove the indexshared state? how that issue is prevented here?